Common/CUDA/Siddon_projection.cu

```diff
@@ -107,6 +107,60 @@ __global__ void vecAddInPlace(float *a, float *b, unsigned long  n)
         a[idx] = a[idx] + b[idx];
 }
 
+template<typename T>
+__device__ T clamp(T low, T high, T value)
+{
+    return (value < low) ? low : ((value > high) ? high : value);
+}
+
+// In a given dimension, determine the number of steps required between entry and exit voxel,
+// and the location of the second voxel intersection.
+// This is used to initialise the Siddon line drawing algorithm.
+template<typename T>
+unsigned long __device__ findLineStepCountAndFirstIntersection(
+    const T rayStart,
+    const T rayEnd,
+    const T rayLength,
+    const T volumeLength,
+    const T alphaMin,
+    const T alphaMax,
+    T& firstIndex,
+    T& step,
+    T& firstAlpha)
+{
+    step = rayStart < rayEnd ? +1 : -1;
+    const T roundDirection = static_cast<T>(+0.5) * static_cast<T>(step);
+    const T minIndex = rayStart < rayEnd ? static_cast<T>(1.0) : static_cast<T>(0.0);
+    const T maxIndex = volumeLength + (minIndex - static_cast<T>(1.0));
+
+    // Find the index of the first "plane" between two voxels
+    // (0 is the leftmost plane, volumeLength is the rightmost plane)
+    // that the ray will cross AFTER having entered the volume (round up if traveling left-to-right).
+    // Then the index of the last "plane" (also round up if traveling left-to-right).
+    const T firstPlane = clamp<T>(
+        minIndex,
+        maxIndex,
+        round(rayStart + alphaMin * rayLength + roundDirection));
+
+    const T lastPlane = clamp<T>(
+        minIndex,
+        maxIndex,
+        round(rayStart + alphaMax * rayLength + roundDirection));
+
+    // Compute the alpha position of the first plane we cross.
+    firstAlpha = (firstPlane - rayStart) / rayLength;
+
+    // Compute the index of the first voxel we hit
+    firstIndex = round(firstPlane - minIndex);
+
+    // If alpha is infinite (= the ray is perpendicular to this axis), make sure it is positive
+    // so it will never be chosen as the minimum.
+    if (isinf(firstAlpha)) { firstAlpha = std::abs(firstAlpha); }
+
+    // Return the number of planes the ray crosses on this axis.
+    return static_cast<unsigned long>(std::abs(lastPlane - firstPlane));

```

Comment:
Code seems mixed with the use of long long, long, and size_t. Would it be safer for this function to return size_t?

Reply:
The original code uses `unsigned long` as well, so I would rather stick to that. You're right that it's a mess of types though. But if we keep the types the same, it will make it easier to merge updates from upstream later :shrug: This isn't super important, because it would only trigger issues with really really big volumes: the number here is the number of voxels which will be crossed by the ray. If this is too large for a 32bit integer, that means the volume X, Y, and/or Z dimensions are each larger than a 32bit integer, which is huge.